### ISSUE TITLE
mssql sorting by primaryKey (duplicate order by column error)

### ISSUE THREAD

**Issue by massivex** [sequelize/sequelize#11258]
## What are you doing?
Querying (through findOne method) a child model with include ordered by primary key

```js
class Project extends Sequelize.Model { }
Project.init({ code: { type: Sequelize.STRING(20) } }, { sequelize });

class Task extends Sequelize.Model { }
Task.init({ name: { type: Sequelize.STRING(20) } }, { sequelize });

Task.belongsTo(Project);

Task.findOne({ include: [ { model: Project } ], order: [['id', 'asc']] });
```
**To Reproduce**
1. Define models like above in code
2. Run query as above (findOne with include and order)
3. See error below
```
SequelizeDatabaseError: A column has been specified more than once in the order by list. Columns in the order by list must be unique.
```

## What do you expect to happen?
A query like that `SELECT [task].[id], [task].[description], [task].[priority] FROM [tasks] AS [task] INNER JOIN [projects] AS [project] ON [task].[projectId] = [project].[id] AND [project].[code] = N'TEST' WHERE [task].[priority] = N'URGENT' ORDER BY [task].[id] ASC OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;`
**with only one occurrence of [task].[id] field in order clause**

## What is actually happening?
Sequelize generates the following query `SELECT [task].[id], [task].[description], [task].[priority] FROM [tasks] AS [task] INNER JOIN [projects] AS [project] ON [task].[projectId] = [project].[id] AND [project].[code] = N'TEST' WHERE [task].[priority] = N'URGENT' ORDER BY [task].[id] ASC, [task].[id] OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;`
**[task].[id] appears two times in order clause**
This happens at line
https://github.com/sequelize/sequelize/blob/c32ac0142032977d968105e400804ed791ee7efb/lib/dialects/mssql/query-generator.js#L872
because SQL Server doesn't allow paging without an order clause. So query generator checks there is an order specified and when it's missing add it.
At line above conditions are satisfied when you're querying from Project (including Task) but not when you're querying from Task (including Project). **So generator adds an extra order clause even if it's defined in options.**
Also, if you specified as order the primary key, SQL SERVER rejects query because it doesn't allow multiple occurrence of the same field in order clause.

I solved updating if with this 
```
if ((!orders.mainQueryOrder || !orders.mainQueryOrder.length) && (!orders.subQueryOrder || !orders.subQueryOrder.length)) {
```
and all these cases work fine
* querying **Task** including Project with **primary key order specified** 
`await Task.findOne({ include: [ { model: Project } ], order: [['id', 'asc']] });`

* querying **Task** including Project with **a non primary key field order specified** 
`await Task.findOne({ include: [ { model: Project } ], order: [['name', 'asc']] });`

* querying **Task** including Project with **no order specified** 
`await Task.findOne({ include: [ { model: Project } ] });`

* querying **Project** including Task with **primary key order specified** 
`await Project.findOne({ include: [ { model: Task } ], order: [['id', 'asc']] });`

* querying **Project** including Task with **a non primary key field order specified** 
`await Project.findOne({ include: [ { model: Task } ], order: [['code', 'asc']] });`

* querying **Project** including Task with **no order specified**
`await Project.findOne({ include: [ { model: Task } ] });`

## Environment
Dialect:
- [ ] mysql
- [ ] postgres
- [ ] sqlite
- [x] mssql
- [ ] any
Dialect **library** version: 5.0.3
Database version: SQL SERVER 2017 - 14.0.3045.24
Sequelize version: 4.44.2
Node Version: 8.11.2
OS: Ubuntu 14
Tested with latest release:
- [ ] No
- [x] Yes, specify that version: 5.12.1

**Comment by papb** [comment]
Thanks for the detailed report! Your solution looks good at a first glance, would you be willing to submit a PR (ideally, adding the six test cases you mentioned as well)? Thanks!

**Comment by massivex** [comment]
Thx, I'll do it

**Comment by massivex** [comment]
Hi @papb ,
I checked out master branch and I found a little issue in tests.

When I run npm tun test (with DIALECT mssql) some tests about transaction stuck.
This is because **MySQL and MSSQL have a different default transaction behaviour**.

When you add/change/remove record linked to a parent model MSSQL, locks both parent & child table rows. Instead MySQL doesn't lock parent row.

For that reason, some tests stuck using mssql dialect failing after 30s by timeout. They fail because transaction tests (after some row insert) try to read from a different transaction newly rows to check if transaction isolation are correctly managed, and without READPAST hist in MSSQL queries hold other transaction completion.

For example this test:
https://github.com/sequelize/sequelize/blob/d041e77e275c3e950cedcdb3a625d891ea8b056b/test/integration/associations/belongs-to-many.test.js#L38

To solve this issue I needed to specify MSSQL table hint "READPAST" and in order to do that in associations test I needed to extend actual tablehint management also for includes (that it's not actually managed).

So, my stuff will be focused on:
- allow tablehint also for include models
- use READPAST tablehint in transaction tests for dialect mssql to avoid test failures and make npm run test-mssql runs correctly
- fix duplicate order by column error

I'm preparing a unique PR with all these changes.

Please, tell me if you prefer to split them into more PRs or if it's fine to merge all them in one PR.

**Comment by papb** [comment]
@massivex Nice work! It's great to see that you found why some tests timeout randomly!!

I think it's better if you split those fixes in more PRs. Hopefully it's not too much extra work.

**Comment by sushantdhiman** [comment]
`npm run setup-mssql` will fix hanging transaction issue if you are following docker guidelines. May need to run that command two times :)

**Comment by dsbert** [comment]
Here is a temporary work-around for anyone waiting for a fix

```js
const qg = require("sequelize/lib/dialects/mssql/query-generator");
// eslint-disable-next-line import/no-extraneous-dependencies
const semver = require("semver");

function addLimitFix(options, model) {
  if (
    semver.valid(this.sequelize.options.databaseVersion) &&
    semver.lt(this.sequelize.options.databaseVersion, "11.0.0")
  ) {
    return "";
  }

  const offset = options.offset || 0;
  const isSubQuery =
    options.subQuery === undefined
      ? options.hasIncludeWhere ||
        options.hasIncludeRequired ||
        options.hasMultiAssociation
      : options.subQuery;

  let fragment = "";
  let orders = {};

  if (options.order) {
    orders = this.getQueryOrders(options, model, isSubQuery);
  }

  if (options.limit || options.offset) {
    if (
      (!orders.mainQueryOrder || !orders.mainQueryOrder.length) &&
      (!orders.subQueryOrder || !orders.subQueryOrder.length)
    ) {
      // if (!options.order || options.include && !orders.subQueryOrder.length) {
      fragment += options.order && !isSubQuery ? ", " : " ORDER BY ";
      fragment += `${this.quoteTable(
        options.tableAs || model.name
      )}.${this.quoteIdentifier(model.primaryKeyField)}`;
    }

    if (options.offset || options.limit) {
      fragment += ` OFFSET ${this.escape(offset)} ROWS`;
    }

    if (options.limit) {
      fragment += ` FETCH NEXT ${this.escape(options.limit)} ROWS ONLY`;
    }
  }

  return fragment;
}

qg.prototype.original_addLimitAndOffset = qg.prototype.addLimitAndOffset;
qg.prototype.addLimitAndOffset = function addLimitAndOffset(options, model) {
  return addLimitFix.call(this, options, model);
};


```

**Comment by papb** [comment]
@dsbert Thanks!!

@massivex Are you still working on this?

**Comment by andrew-st-angelo-77media** [comment]
I'm in the process of upgrading our Sequelize implementation.  I can confirm I'm seeing this is v4.44.  Is this affecting v5, as well?

**Comment by massivex** [comment]
> I'm in the process of upgrading our Sequelize implementation. I can confirm I'm seeing this is v4.44. Is this affecting v5, as well?

yes v5 have these issues and some others.

I don't know if last changes fix one or all these but I attached all my stuff I did to make last sequelize v5 release works fine with MSSQL. 


[sequelize.zip](https://github.com/sequelize/sequelize/files/4350071/sequelize.zip)

zip attached solves these issues and it's provided without any warranties. It's written in Typescript.
I hope that can help you. I hope also others can create a pull request because I didn't find time to do it.

**Solved issues**
1. fix addLimitAndOffset method to avoid duplicate id field in order
2. fix upsertQuery to escape correctly all values according with model settings 
3. fix upsertQuery to fix unique index detection. It does not work when clause[key] is falsy (for example for numbers equals to 0). I replaced condition to force null or undefined match.
4. fix DATEONLY I patched sanitize function to return a string instead a Date object as sequelize does in MYSQL dialect. I did it because DATEONLY should not be timezone aware and Date object it's timezone aware.
5. added VARCHAR data type. By default all strings in sequelize are unicode and inside where clauses are translated with the N character prefix (N prefix it's used for specify unicode encoded string). VARCHAR fields doesn't support unicode string and often when you prefix with N values for non unicode fields mssql doesn't build the correct execution plan. That means changes between query execution from 0.1 seconds to 10 seconds (depends on how are large your tables)
6. added GEOMETRY. I wrote code to support GEOMETRY in MSSQL. A very basic implementation. It's very similar to MYSQL implentation... I changed only parsing function to parse correctly geometry binary object returned by MSSQL
7. fix DECIMAL to resolve an issue related to how tedious works with decimals. When the number is not formatted in the same scale as field definition, tedious wrote a number in database that doesn't match the value that you provided. In that case sequelize logs returns the correct value but database store a wrong number!?! I found this issue tracing MSSQL with SQL Server Profiler

**How to use**
```
// import functions
import { sequelizeAdditions, patchSequelize } from "./sequelize";

// execute that before creating sequelize object
patchSequelize(); // this method fix issues overwriting methods

// ... create sequelize object ...

sequelizeAdditions(); // this method add and fix data types 

// force refreshDataTypes to load new data types
const DataTypes = require("sequelize/lib/data-types").mssql;
sequelize.dialect.connectionManager.refreshTypeParser(DataTypes);

```
